plugins/modules/identity/keycloak/keycloak_user.py

```diff
@@ -0,0 +1,527 @@
+        new_param_value = module.params.get(param)
+        old_value = before_user[param] if param in before_user else None
+        if new_param_value != old_value:
+            changeset[camel(param)] = new_param_value
+
+    # Prepare the desired values using the existing values (non-existence results in a dict that is save to use as a basis)
+    desired_user = before_user.copy()
+    desired_user.update(changeset)
+
+    # Cater for when it doesn't exist (an empty dict)
+    if not before_user:
+        if state == 'absent':
+            # Do nothing and exit
+            if module._diff:
+                result['diff'] = dict(before='', after='')
+            result['changed'] = False
+            result['end_state'] = {}
+            result['user'] = result['end_state']
+            result['msg'] = 'User does not exist; doing nothing.'
+            module.exit_json(**result)
+
+        # Process a creation
+        result['changed'] = True
+
+        if module._diff:
+            result['diff'] = dict(before='', after=desired_user)
+
+        if module.check_mode:
+            module.exit_json(**result)
+
+        # create it
+        kc.create_user(desired_user, realm=realm)
+
+        after_user = kc.get_users_by_username(username, realm=realm)[0]
+
+        result['end_state'] = after_user
+        result['user'] = result['end_state']
+
+        result['msg'] = 'User {username} has been created with ID {id}'.format(username=after_user['username'],
+                                                                               id=after_user['id'])
+        module.exit_json(**result)
+
+    else:
+        if state == 'present':
+            # Process an update
+
+            # no changes
+            if desired_user == before_user:
+                result['changed'] = False
+                result['end_state'] = desired_user
+                result['user'] = result['end_state']
+                result['msg'] = "No changes required to user {username}.".format(
+                    username=before_user['username'])
+                module.exit_json(**result)
+
+            # doing an update
+            result['changed'] = True
+
+            if module._diff:
+                result['diff'] = dict(before=before_user, after=desired_user)
+
+            if module.check_mode:
+                module.exit_json(**result)
+
+            # do the update
+            kc.update_user(desired_user, realm=realm)
+
+            after_user = kc.get_users_by_username(username, realm=realm)[0]
+
+            result['end_state'] = after_user
+            result['user'] = result['end_state']
+
+            result['msg'] = "User {id} has been updated".format(
+                id=after_user['id'])
+            module.exit_json(**result)
+
+        else:
+            # Process a deletion (because state was not 'present')
+            result['changed'] = True
+
+            if module._diff:
+                result['diff'] = dict(before=before_user, after='')
```

Comment:
Given that the `before_user` is a dict, I think it's best if the `after` is also a dict rather than an empty string.